src/function/gds/single_shortest_paths.cpp

```diff
@@ -146,12 +147,12 @@ class SingleSPOutputWriterPaths : public SingleSPOutputWriterLengths {
         lengthVector->setValue<int64_t>(0, length);
         PathVectorWriter writer(pathNodeIDsVector.get());
         writer.beginWritingNewPath(length);
-        common::nodeID_t curIntNode = dstNodeID;
-        while (writer.nextPathPos > 0) {
-            curIntNode =
-                rjOutputs->ptrCast<SingleSPOutputsPaths>()->singlePaths->getParent(curIntNode);
-            writer.addNewNodeID(curIntNode);
+        common::nodeID_t nodeID = dstNodeID;
+        for (auto i = 0u; i < length - 1; ++i) {
+            nodeID = rjOutputs->ptrCast<SingleSPOutputsPaths>()->singlePaths->getParent(nodeID);
+            writer.addNode(nodeID, length - 2 - i);
```

Comment:
Single SPs is still not returning edgeIDs. Just so we are on the same page: one design I would consider is to remove SingleSP function completely and give a bool parameter "isSinglePaths" to all_sp_paths (since it will be using the same algorithm). But then use the parameter to change this block to a single cas operation instead of a while loop (if the first cas fails, it's OK, because there is another winner. That is we can skip this loop: 
```
while (!curPtr[childNodeID.offset].compare_exchange_strong(expected, parentEdgePtr))
            ;
```
I would similarly optimize the output DSF writer code (e.g., the iter-1 search that I commented about inside the while loop is not needed). Or you might still want to keep a separate outputwriter but at least the computation I think can be completely shared.

Reply:
Yeah I'm making this change in the coming PR.